src/OSPSuite.Core/Domain/Builder/SimulationBuilder.cs

```diff
@@ -86,16 +87,43 @@ internal IEnumerable<InitialCondition> AllPresentMoleculeValuesFor(IEnumerable<s
 
       private void performMerge()
       {
-         _passiveTransports.AddRange(allBuilder(x => x.PassiveTransports));
+         cacheMoleculeDependentBuilders(x => x.PassiveTransports, _passiveTransports);
+
          _reactions.AddRange(allBuilder(x => x.Reactions));
          _eventGroups.AddRange(allBuilder(x => x.EventGroups));
-         _observers.AddRange(allBuilder(x => x.Observers));
+
+         cacheMoleculeDependentBuilders(x => x.Observers, _observers);
+
          _molecules.AddRange(allBuilder(x => x.Molecules));
          _parameterValues.AddRange(allStartValueBuilder(x => x.SelectedParameterValues));
          _initialConditions.AddRange(allStartValueBuilder(x => x.SelectedInitialConditions)
             .Concat(_simulationConfiguration.ExpressionProfiles.SelectMany(x => x.InitialConditions)));
       }
 
+      private void cacheMoleculeDependentBuilders<T>(Func<Module, IBuildingBlock<T>> propAccess, ObjectBaseCache<T> cache) where T : class, IMoleculeDependentBuilder
+      {
+         var builders = allBuilder(propAccess).ToList();
+         cache.AddRange(builders);
+         cacheMoleculeLists(builders, cache);
+      }
+
+      private void cacheMoleculeLists<T>(IReadOnlyList<T> allBuilders, ObjectBaseCache<T> builderCache) where T : class, IMoleculeDependentBuilder
+      {
+         builderCache.Each(builder => { combineMoleculeLists(allBuilders.Where(x => string.Equals(x.Name, builder.Name)).ToList(), builder); });
```

Comment:
that's super hard to read here. The variable of allBuilders is called x but we iterate over a. variable call builder . I don't know what's going on
ToList not required here probably

Reply:
Maybe rename builder to something else, like builderUsedInSimulation, lastBuilder, etc.